File: annotation-processor/src/main/java/me/bristermitten/mittenlib/annotations/parser/CustomSerializers.java
package me.bristermitten.mittenlib.annotations.parser;

import com.google.inject.Singleton;
import com.squareup.javapoet.ClassName;
import io.toolisticon.aptk.tools.MessagerUtils;
import io.toolisticon.aptk.tools.TypeUtils;
import io.toolisticon.aptk.tools.wrapper.ExecutableElementWrapper;
import io.toolisticon.aptk.tools.wrapper.TypeElementWrapper;
import me.bristermitten.mittenlib.annotations.ast.CustomSerializerInfo;
import me.bristermitten.mittenlib.config.SerializationContext;
import me.bristermitten.mittenlib.config.extension.CustomSerializer;
import me.bristermitten.mittenlib.config.extension.CustomSerializerFor;
import me.bristermitten.mittenlib.config.tree.DataTree;

import javax.lang.model.element.Modifier;
import javax.lang.model.element.TypeElement;
import javax.lang.model.type.MirroredTypeException;
import javax.lang.model.type.TypeMirror;
import java.util.Optional;

@Singleton
public class CustomSerializers extends CustomInfoRegistry<CustomSerializerInfo> {

    public void registerCustomSerializer(TypeElement customSerializerType) {
        CustomSerializerFor annotation = customSerializerType.getAnnotation(CustomSerializerFor.class);
        if (annotation == null) {
            throw new IllegalArgumentException(
                    "CustomSerializer must be annotated with @CustomSerializerFor");
        }

        TypeMirror serializerFor;
        try {
            var ignored = annotation.value();
            throw new IllegalStateException("Expected MirroredTypeException");
        } catch (MirroredTypeException e) {
            serializerFor = e.getTypeMirror();
        }

        var implementsCustomSerializer =
                TypeElementWrapper.wrap(customSerializerType).getAllInterfaces().stream()
                        .anyMatch(i -> i.getQualifiedName().equals(CustomSerializer.class.getCanonicalName()));

        Optional<ExecutableElementWrapper> serializeMethodOpt =
                TypeElementWrapper.wrap(customSerializerType).getMethods().stream()
                        .filter(m -> m.getSimpleName().equals("serialize"))
                        .filter(m -> m.getParameters().size() == 2)
                        .filter(
                                m ->
                                        TypeUtils.TypeComparison.isTypeEqual(
                                                m.getParameters().getFirst().asType().unwrap(), serializerFor))
                        .filter(
                                m ->
                                        m.getParameters()
                                                .get(1)
                                                .asType()
                                                .toString()
                                                .equals(SerializationContext.class.getCanonicalName()))
                        .findFirst();

        if (serializeMethodOpt.isPresent()) {
            var method = serializeMethodOpt.get();
            if (!method.getReturnType().unwrap().toString().equals(DataTree.class.getCanonicalName())) {
                MessagerUtils.error(method.unwrap(), "Custom serializer method must return DataTree");
                return;
            }
        }

        if (!implementsCustomSerializer && serializeMethodOpt.isEmpty()) {
            throw new IllegalArgumentException(
                    "CustomSerializer must implement CustomSerializer or have a static method DataTree serialize(T, SerializationContext)");
        }

        boolean isStatic =
                !implementsCustomSerializer
                        && serializeMethodOpt.isPresent()
                        && serializeMethodOpt.get().unwrap().getModifiers().contains(Modifier.STATIC);

        if (!isStatic && !implementsCustomSerializer) {
            MessagerUtils.error(
                    customSerializerType, "Non static custom serializers must implement CustomSerializer");
            return;
        }

        var customSerializerInfo = new CustomSerializerInfo(customSerializerType, isStatic);

        register(ClassName.get(serializerFor), customSerializerInfo);
    }
}

File: annotation-processor/src/main/scala/me/bristermitten/mittenlib/annotations/parser/CustomSerializers.scala
package me.bristermitten.mittenlib.annotations.parser

import com.google.inject.Singleton
import com.palantir.javapoet.TypeName
import io.toolisticon.aptk.tools.MessagerUtils
import io.toolisticon.aptk.tools.TypeUtils
import io.toolisticon.aptk.tools.wrapper.ExecutableElementWrapper
import io.toolisticon.aptk.tools.wrapper.TypeElementWrapper
import java.util.Optional
import javax.lang.model.element.Modifier
import javax.lang.model.element.TypeElement
import javax.lang.model.`type`.MirroredTypeException
import javax.lang.model.`type`.TypeMirror
import me.bristermitten.mittenlib.annotations.ast.CustomSerializerInfo
import me.bristermitten.mittenlib.config.SerializationContext
import me.bristermitten.mittenlib.config.extension.CustomSerializer
import me.bristermitten.mittenlib.config.extension.CustomSerializerFor
import me.bristermitten.mittenlib.config.tree.DataTree
import scala.jdk.CollectionConverters.*

@Singleton
class CustomSerializers extends CustomInfoRegistry[CustomSerializerInfo]:

  def registerCustomSerializer(customSerializerType: TypeElement): Unit =
    val annotation =
      customSerializerType.getAnnotation(classOf[CustomSerializerFor])
    if (annotation == null) {
      MessagerUtils.error(
        customSerializerType,
        "CustomSerializer must be annotated with @CustomSerializerFor"
      )
      return
    }

    val serializerFor =
      try
        val _ = annotation.value()
        throw new IllegalStateException("Expected MirroredTypeException")
      catch case e: MirroredTypeException => e.getTypeMirror

    val implementsCustomSerializer = TypeElementWrapper
      .wrap(customSerializerType)
      .getAllInterfaces
      .asScala
      .exists(i =>
        i.getQualifiedName.equals(classOf[CustomSerializer[?]].getCanonicalName)
      )

    val serializeMethodOpt = TypeElementWrapper
      .wrap(customSerializerType)
      .getMethods()
      .asScala
      .find { m =>
        m.getSimpleName == "serialize" &&
        m.getParameters.size() == 2 &&
        TypeUtils.TypeComparison.isTypeEqual(
          m.getParameters.get(0).asType().unwrap(),
          serializerFor
        ) &&
        m.getParameters
          .get(1)
          .asType()
          .toString == classOf[SerializationContext].getCanonicalName
      }

    if (serializeMethodOpt.isDefined) {
      val method = serializeMethodOpt.get
      if (
        method.getReturnType
          .unwrap()
          .toString != classOf[DataTree].getCanonicalName
      ) {
        MessagerUtils.error(
          method.unwrap(),
          "Custom serializer method must return DataTree"
        )
        return
      }
    }

    if (!implementsCustomSerializer && serializeMethodOpt.isEmpty) {
      MessagerUtils.error(
        customSerializerType,
        "CustomSerializer must implement CustomSerializer or have a static method DataTree serialize(T, SerializationContext)"
      )
      return
    }

    val isStatic = !implementsCustomSerializer &&
      serializeMethodOpt.isDefined &&
      serializeMethodOpt.get.unwrap().getModifiers.contains(Modifier.STATIC)

    if (!isStatic && !implementsCustomSerializer) {
      MessagerUtils.error(
        customSerializerType,
        "Non static custom serializers must implement CustomSerializer"
      )
      return
    }

    val customSerializerInfo =
      CustomSerializerInfo(customSerializerType, isStatic)

    register(TypeName.get(serializerFor), customSerializerInfo)
